qendpoint-store/src/test/java/com/the_qa_company/qendpoint/TempTest.java

```diff
@@ -0,0 +1,189 @@
+
+		Path hdtstore = root.resolve("hdt-store");
+		Path locationNative = root.resolve("native");
+
+		Files.createDirectories(hdtstore);
+		Files.createDirectories(locationNative);
+
+		String indexName = "index.hdt";
+
+		HDTOptions options = HDTOptions.of(
+				// disable the default index (to use the custom indexes)
+				HDTOptionsKeys.BITMAPTRIPLES_INDEX_NO_FOQ, true,
+				// set the custom indexes we want
+				HDTOptionsKeys.BITMAPTRIPLES_INDEX_OTHERS, "sop,ops,osp,pso,pos");
+
+
+		try (HDT hdt = HDTManager.generateHDT(new Iterator<>() {
+			@Override
+			public boolean hasNext() {
+				return false;
+			}
+
+			@Override
+			public TripleString next() {
+				return null;
+			}
+		}, Utility.EXAMPLE_NAMESPACE, options, null)) {
+			hdt.saveToHDT(hdtstore.resolve(indexName).toAbsolutePath().toString(), null);
+		} catch (Error | RuntimeException e) {
+			throw e;
+		} catch (Exception e) {
+			throw new RuntimeException(e);
+		}
+
+		repository = CompiledSail.compiler().withEndpointFiles(new EndpointFiles(locationNative, hdtstore, indexName))
+				.compileToSparqlRepository();
+		try (InputStream is = new BufferedInputStream(Objects.requireNonNull(loader.getResourceAsStream(filename),
+				filename + " doesn't exist"))) {
+			repository.loadFile(is, filename);
+		}
+	}
+
+	@After
+	public void after() {
+		if (repository != null) {
+			repository.shutDown();
+		}
+		repository = null;
+	}
+
+	@Test
+	public void test() {
+		try (SailRepositoryConnection connection = repository.getConnection()) {
+			System.out.println();
+			String query = """
+					PREFIX epo: <http://data.europa.eu/a4g/ontology#>
+					PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+					PREFIX legal: <https://www.w3.org/ns/legal#>
+					PREFIX dcterms: <http://purl.org/dc/terms#>
+					PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
+					PREFIX dc: <http://purl.org/dc/elements/1.1/>
+					      
+					SELECT DISTINCT ?countryID ?year (COUNT(DISTINCT ?lot) AS ?amountLots) (SUM(if(?bidders = 1, 1, 0)) AS ?numSingleBidders) WHERE {
+					      
+					        ?proc a epo:Procedure .
+					        ?proc epo:hasProcedureType ?p .
+					        ?proc epo:hasProcurementScopeDividedIntoLot ?lot .
+					      
+					      	?stat epo:concernsSubmissionsForLot ?lot .
+					      
+					        ?stat a epo:SubmissionStatisticalInformation .
+					        ?stat epo:hasReceivedTenders ?bidders .
+					      	
+					      	?resultnotice epo:refersToProcedure ?proc .
+					        ?resultnotice a epo:ResultNotice .
+					        ?resultnotice epo:hasDispatchDate ?ddate .
+					      
+					      	FILTER ( ?p != <http://publications.europa.eu/resource/authority/procurement-procedure-type/neg-wo-call>)
+							BIND(year(xsd:dateTime(?ddate)) AS ?year) .
+								
+							# This line should be above the FILTER and BIND, but doing so causes the QueryJoinOptimizer to not optimize the query for some unknown reason	
+					        ?resultnotice epo:refersToRole ?buyerrole .				      	
```

Comment:
@D063520 @ate47 

This line should be above the FILTER and BIND. When I put it above then the QueryJoinOptimizer doesn't manage to optimize the query at all. When I move it back down again like I've done now then it does end up triggering the QueryJoinOptimizer seemingly correctly, but it only seems to check the cardinality of this single statement pattern here and when it does it gets 0.0 as the cardinality.

If you run the test you should see the following printed:

```
Cardinality for StatementPattern    Var (name=resultnotice)    Var (name=_const_6aa9a9c_uri, value=http://data.europa.eu/a4g/ontology#refersToRole, anonymous)    Var (name=buyerrole)  is 0.0 (HDT: 0.0, Native: 0.0)
```

Reply:
I had a look at the QueryJoinOptimizer code and it doesn't seem to recurse down into the query tree. Essentially it will recurse down to the first `Join` and then optimize and optimize and join arguments that are either instance of `Join` or `LeftJoin` or `StatementPattern`.

Reply:
I fixed the QueryJoinOptimizer https://github.com/eclipse-rdf4j/rdf4j/pull/4879